### PR TITLE
OTFMapFusion: Support for WCR, nested maps & connector renaming fixed

### PR DESCRIPTION
**Pull request by lukastruemper**
This PR extends and fixes OTFMapFusion

- Fusion of maps with WCR in second map
- Fusion of maps with NestedSDFGs
- Renaming of symbols to avoid symbol collision of map parameters
- Renaming of symbols in NestedSDFG to avoid collision with map parameters

### CODE REVIEW

File: dace/transformation/dataflow/otf_map_fusion.py [hunk 1]
```diff
@@ -50,44 +49,93 @@ def can_be_applied(self, graph, expr_index, sdfg, permissive=False):
...
+                    return False
 
-            write_memlets[array] = memlet
+                length = symbolic.evaluate(e - b, symbols=sdfg.constants)
+                if length > 1:
+                    return False
 
-        # Memlets access must be inferable
-        first_map_entry = graph.entry_node(self.first_map_exit)
+            # WCR must have identity zero. Otherwise, we must specifically initialize the temporary data
```
Comment (tbennun) [inline dace/transformation/dataflow/otf_map_fusion.py:74]: that is also the case for zero identity, but you use setzero. Please note in the code that you use `setzero` below.

File: dace/transformation/dataflow/otf_map_fusion.py [hunk 2]
```diff
@@ -50,44 +49,93 @@ def can_be_applied(self, graph, expr_index, sdfg, permissive=False):
...
-                return False
+            # Only fuse scalars
+            for access in memlet.subset:
+                b, e, s = access
+                if s > 1:
+                    return False
+
+                length = symbolic.evaluate(e - b, symbols=sdfg.constants)
+                if length > 1:
+                    return False
```
Comment (tbennun) [inline dace/transformation/dataflow/otf_map_fusion.py:110]: same as above.

File: dace/transformation/dataflow/otf_map_fusion.py [hunk 3]
```diff
@@ -96,102 +144,185 @@ def apply(self, graph: SDFGState, sdfg: SDFG):
...
             else:
-                raise ValueError("Failed to connect")
+                suffix = "OTF"
+
+            new_in_connector = "IN_" + suffix
+            if not self.second_map_entry.add_in_connector(new_in_connector):
+                for n in range(2, 65536):
+                    new_in_connector = "IN_" + suffix + "_" + str(n)
+                    if self.second_map_entry.add_in_connector(new_in_connector):
+                        break
```
Comment (tbennun) [inline dace/transformation/dataflow/otf_map_fusion.py:166]: I still don't understand why you do not use `next_connector`

File: dace/transformation/dataflow/otf_map_fusion.py [hunk 4]
```diff
@@ -96,102 +144,185 @@ def apply(self, graph: SDFGState, sdfg: SDFG):
...
+                    node.map.params = inner_map_params
+
+                    # Substitue range definitions by symbols for second map
+                    ranges = []
+                    for access in node.map.range:
+                        b, e, s = access
+
+                        b = symbolic.pystr_to_symbolic(b)
+                        e = symbolic.pystr_to_symbolic(e)
+                        s = symbolic.pystr_to_symbolic(s)
```
Comment (tbennun) [inline dace/transformation/dataflow/otf_map_fusion.py:260]: they should already be symbolic, no?

File: dace/transformation/dataflow/otf_map_fusion.py [hunk 5]
```diff
@@ -96,102 +144,185 @@ def apply(self, graph: SDFGState, sdfg: SDFG):
...
+
+                        b = b.subs(param_mapping)
+                        e = e.subs(param_mapping)
+                        s = s.subs(param_mapping)
+
+                        ranges.append((str(b), str(e), str(s)))
```
Comment (tbennun) [inline dace/transformation/dataflow/otf_map_fusion.py:266]: oh no, please do not use strings if you do not have to

File: dace/transformation/dataflow/otf_map_fusion.py [hunk 6]
```diff
@@ -96,102 +144,185 @@ def apply(self, graph: SDFGState, sdfg: SDFG):
...
+                            b, e, s = access
+                            b = symbolic.pystr_to_symbolic(b)
+                            e = symbolic.pystr_to_symbolic(e)
+                            s = symbolic.pystr_to_symbolic(s)
+
+                            b = b.subs(scope_subs)
+                            e = e.subs(scope_subs)
+                            s = s.subs(scope_subs)
+
+                            ranges.append((str(b), str(e), str(s)))
```
Comment (tbennun) [inline dace/transformation/dataflow/otf_map_fusion.py:286]: same

File: dace/transformation/dataflow/otf_map_fusion.py [hunk 7]
```diff
@@ -96,102 +144,185 @@ def apply(self, graph: SDFGState, sdfg: SDFG):
...
+                                                   self.second_map_entry.map.params, in_accesses)
+
+                # Copy actual content
+                first_map_inner_nodes = self._copy_first_map_contents(sdfg, graph, first_map_entry)
+
+                # Prepare: Symbol collision
+                # Second map defines symbols (params) and becomes the new outer scope of the copied nodes.
+                # Thus, newly introduced symbols in this subgraph must be renamed first to avoid conflicts.
+                # Note symbols of first map must be re-named according to the param mapping.
+                for node in first_map_inner_nodes:
```
Comment (tbennun) [inline dace/transformation/dataflow/otf_map_fusion.py:232]: this won't work for multi-level nested maps, right?

File: dace/transformation/dataflow/otf_map_fusion.py [hunk 8]
```diff
@@ -96,102 +144,185 @@ def apply(self, graph: SDFGState, sdfg: SDFG):
...
 
-                            for param, sub in param_subs[i].items():
-                                b = b.subs(param, sub)
-                                e = e.subs(param, sub)
-                                s = s.subs(param, sub)
+                            b = b.subs(param_mapping)
+                            e = e.subs(param_mapping)
+                            s = s.subs(param_mapping)
 
                             ranges.append((str(b), str(e), str(s)))
```
Comment (tbennun) [inline dace/transformation/dataflow/otf_map_fusion.py:327]: same

File: dace/transformation/dataflow/otf_map_fusion.py [hunk 9]
```diff
@@ -322,27 +454,40 @@ def solve(first_params, write_accesses, second_params, read_accesses):
             # Step is constant
             assert (s0 - s1) == 0
 
+            b_eq = sympy.Eq(b0, b1)
+            e_eq = sympy.Eq(e0, e1)
```
Comment (tbennun) [inline dace/transformation/dataflow/otf_map_fusion.py:458]: used outside the `if`?

**Review by tbennun — CHANGES_REQUESTED** [review]
I'm a bit worried about the symbol replacement

File: dace/transformation/dataflow/otf_map_fusion.py (hunk 3 above)
Comment (lukastruemper) [reply dace/transformation/dataflow/otf_map_fusion.py:166]: Yeah, super annoying. The solution is awkward, but calling "while add_in_connector == False: new_connector = next_connector(new_connector)" was causing problems. I couldn't debug them, but can try another time. I know that it occurs in hdiff when fusing all maps

File: dace/transformation/dataflow/otf_map_fusion.py (hunk 7 above)
Comment (lukastruemper) [reply dace/transformation/dataflow/otf_map_fusion.py:232]: No, basically everyhting nested, also no states, nested SDFGs etc. Here I would need your help to generalize it. There must be a helper function which could replace symbols in subgraphs?

**Comment by lukastruemper** [comment]
> I'm a bit worried about the symbol replacement

I addressed the simple problems now, but as we discussed, the symbol replacement needs to be handled better. Maybe you could show me how to make a nice (and more general) solution?

File: dace/transformation/dataflow/otf_map_fusion.py (hunk 3 above)
Comment (lukastruemper) [reply dace/transformation/dataflow/otf_map_fusion.py:154]: I will look into this separately

File: dace/transformation/dataflow/otf_map_fusion.py (hunk 7 above)
Comment (tbennun) [reply dace/transformation/dataflow/otf_map_fusion.py:220]: `SubgraphView.replace` is your friend

File: dace/sdfg/replace.py [hunk 10]
```diff
@@ -60,7 +60,10 @@ def replace_dict(subgraph: 'dace.sdfg.state.StateGraphView',
 
     # Replace in node properties
     for node in subgraph.nodes():
-        replace_properties_dict(node, repl, symrepl)
+        if isinstance(node, dace.nodes.NestedSDFG):
+            node.sdfg.replace_dict(repl, symrepl)
```
Comment (lukastruemper) [inline dace/sdfg/replace.py:64]: @tbennun This causes many other tests to fail, is this a bad idea?

File: dace/sdfg/replace.py [hunk 11]
```diff
@@ -148,6 +151,14 @@ def replace_properties_dict(node: Any,
                     propval[symname] = symbolic.pystr_to_symbolic(str(sym_mapping)).subs(symrepl)
                 except AttributeError:  # If the symbolified value has no subs
                     pass
+        elif isinstance(propclass, properties.ListProperty) and pname == 'params':
```
Comment (lukastruemper) [inline dace/sdfg/replace.py:154]: @tbennun This seems to be a bug: When renaming a parameter of a map, the parameters gets renamed correctly in the scope of the map (memlets, etc) but doesn't get renamed in the map definition.

If I add these lines, many of our tests start to break

File: dace/transformation/dataflow/otf_map_fusion.py [hunk 12]
```diff
@@ -50,44 +49,98 @@ def can_be_applied(self, graph, expr_index, sdfg, permissive=False):
...
+                return False
+            first_memlets[memlet.data] = memlet
+
+        second_memlets = {}
         for edge in graph.out_edges(self.second_map_entry):
-            read_memlet = edge.data
-            if read_memlet.data not in write_memlets:
+            memlet = edge.data
+
+            if not memlet.data in first_memlets:
```
Comment (tbennun) [inline dace/transformation/dataflow/otf_map_fusion.py:93]: same as above

**Review by tbennun — CHANGES_REQUESTED** [review]
Only minor comments remain, otherwise LGTM!

File: dace/transformation/dataflow/otf_map_fusion.py [hunk 13]
```diff
@@ -50,44 +49,98 @@ def can_be_applied(self, graph, expr_index, sdfg, permissive=False):
...
+                red_type = detect_reduction_type(memlet.wcr)
+                if red_type == dtypes.ReductionType.Custom:
+                    return False
 
-        # Memlets access must be inferable
-        first_map_entry = graph.entry_node(self.first_map_exit)
+                dtype = sdfg.arrays[memlet.data].dtype
+                identity = dtypes.reduction_identity(dtype, red_type)
+                if identity != 0:
+                    return False
```
[the remark this answers is not in the document]
Comment (lukastruemper) [reply dace/transformation/dataflow/otf_map_fusion.py:77]: Yes, otherwise we cannot use the setzero on the temporary buffer for initialization.

**Comment by lukastruemper** [comment]
> Only minor comments remain, otherwise LGTM!

Pushed the fixes

File: dace/transformation/dataflow/otf_map_fusion.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (lukastruemper) [reply dace/transformation/dataflow/otf_map_fusion.py:77]: In an OTF v2, we could insert a NestedSDFG which initializes the temporary buffer first

**Review by tbennun — APPROVED** [review]
LGTM!

**Pull request merged**